### PR TITLE
Cofunction coarsening

### PR DESCRIPTION
**Pull request by nbouziani**
Add support for coarsening of cofunctions, and type checking for `prolong`, `restrict`, `inject`.

fixes https://github.com/firedrakeproject/firedrake/issues/3169

### CODE REVIEW

File: firedrake/mg/kernels.py [hunk 1]
```diff
@@ -310,7 +311,8 @@ def restrict_kernel(Vf, Vc):
         return cache[key]
     except KeyError:
         mesh = extract_unique_domain(coordinates)
-        evaluate_code = compile_element(firedrake.TestFunction(Vc), Vf)
+        v = firedrake.TestFunction(Vc.dual()) if isinstance(Vc, FiredrakeDualSpace) else firedrake.TestFunction(Vc)
+        evaluate_code = compile_element(v, Vf)
```
Comment (nbouziani) [inline firedrake/mg/kernels.py:315]: I am not entirely sure this is correct.

Comment (pbrubeck) [reply firedrake/mg/kernels.py:315]: At this point we are guaranteed that `Vc` is a `DualSpace`.

Comment (pbrubeck) [reply firedrake/mg/kernels.py:315]: Oh, this is not the case for more than 1 refinement per level.

File: firedrake/mg/ufl_utils.py [hunk 2]
```diff
@@ -104,6 +105,12 @@ def coarsen_form(form, self, coefficient_mapping=None):
     return form
 
 
+@coarsen.register(ufl.FormSum)
+def coarsen_formsum(form, self, coefficient_mapping=None):
+    return type(form)(*[(self(i, self, coefficient_mapping=coefficient_mapping), wi)
```
Comment (pbrubeck) [inline firedrake/mg/ufl_utils.py:110]: Should we also coarsen the weights?

**Review by dham — CHANGES_REQUESTED** [review]
Apart from fixing the ugly cache, this is fine.

File: firedrake/mg/ufl_utils.py (hunk 2 above)
Comment (pbrubeck) [reply firedrake/mg/ufl_utils.py]: At the moment weights are always constant, but here we generalize any future extensions

File: firedrake/mg/embedded.py [hunk 3]
```diff
@@ -173,11 +173,16 @@ def DG_work(self, V):
         :arg V: a function space.
         :returns: A Function in the embedding DG space.
         """
+        needs_dual = ufl.duals.is_dual(V)
         cache = self.cache(V.ufl_element())
-        key = V.dim()
+        key = (V.dim(), needs_dual)
```
Comment (pbrubeck) [inline firedrake/mg/embedded.py:178]: @dham This should address point (1/2) from the review. I used recursion to populate both Function and Cofunction caches.

File: firedrake/mg/interface.py [hunk 4]
```diff
@@ -3,22 +3,29 @@
...
 from . import utils
 from . import kernels
 
 
 __all__ = ["prolong", "restrict", "inject"]
 
 
-def check_arguments(coarse, fine):
+def check_arguments(coarse, fine, needs_dual=False):
+    if is_dual(coarse) != needs_dual:
```
Comment (pbrubeck) [inline firedrake/mg/interface.py:15]: @dham This should address point (2/2) from the review. Now we are type checking using `is_dual`.

File: tests/multigrid/test_poisson_gmg.py [hunk 5]
```diff
@@ -113,3 +117,42 @@ def test_poisson_gmg_parallel_fas():
...
+               ]
+    bcs.restore()
+    solutions = []
+    for L in sources:
+        uh = Function(V)
+        solve(a == L, uh, bcs=bcs, solver_parameters=parameters)
+        solutions.append(uh)
+
+    for s in solutions[1:]:
+        assert np.allclose(s.dat.data_ro, solutions[0].dat.data_ro, rtol=1E-14)
```
Comment (JHopeCollins) [inline tests/multigrid/test_poisson_gmg.py:158]: Why not just `assert errornorm(s, solutions[0]) < 1e-14)`?

**Comment by JHopeCollins** [comment]
I'm happy that this fixes my MFE in #3169, and it looks like that case is covered in the tests.

**Review by JHopeCollins — APPROVED** [review]

File: tests/multigrid/test_poisson_gmg.py (hunk 5 above)
Comment (pbrubeck) [reply tests/multigrid/test_poisson_gmg.py:158]: This should not generate code

File: firedrake/mg/ufl_utils.py [hunk 6]
```diff
@@ -160,17 +172,23 @@ def coarsen_function_space(V, self, coefficient_mapping=None):
...
     if new is None:
-        V = expr.function_space()
-        manager = firedrake.dmhooks.get_transfer_manager(expr.function_space().dm)
-        V = self(expr.function_space(), self)
-        new = firedrake.Function(V, name="coarse_%s" % expr.name())
-        manager.inject(expr, new)
+        Vf = expr.function_space()
+        manager = firedrake.dmhooks.get_transfer_manager(Vf.dm)
+        Vc = self(Vf, self)
+        new = firedrake.Function(Vc, name="coarse_%s" % expr.name())
```
Comment (connorjward) [inline firedrake/mg/ufl_utils.py:185]:
```suggestion
        new = firedrake.Function(Vc, name=f"coarse_{expr.name()}")
```

File: tests/multigrid/test_poisson_gmg.py (hunk 5 above)
Comment (connorjward) [reply tests/multigrid/test_poisson_gmg.py:158]: Why does that matter? It definitely reads better to have `errornorm`.

**Review by connorjward — CHANGES_REQUESTED** [review]
This is just about ready. Address these and I'll approve it.

File: tests/multigrid/test_poisson_gmg.py (hunk 5 above)
Comment (pbrubeck) [inline tests/multigrid/test_poisson_gmg.py:158]:
```suggestion
        assert errornorm(s, solutions[0]) < 1E-14
```

**Review by connorjward — APPROVED** [review]

**Pull request merged**